src/librustc_typeck/astconv.rs

```diff
@@ -1731,8 +1731,14 @@ pub fn partition_bounds<'a>(tcx: &ty::ctxt,
                         if ty::try_add_builtin_trait(tcx,
                                                      trait_did,
                                                      &mut builtin_bounds) {
-                            // FIXME(#20302) -- we should check for things like Copy<T>
-                            continue; // success
+                            let segments = &b.trait_ref.path.segments;
+                            let parameters = &segments[segments.len() - 1].parameters;
+                            if parameters.is_empty() {
+                                continue; // success
+                            }
+                            span_fatal!(tcx.sess, b.trait_ref.path.span, E0316,
+                                        "builtin bounds do not require type arguments, {} given",
+                                        parameters.types().len());
```

Comment:
@ahmedcharles similar tests to the ones you have should be ok. For example:

`struct MyStruct<'a, T: Copy<'a>>;`

`fn foo<'a, T:Copy<'a, U>, U>(x: T) {}`

Trait implementations are checked correctly, the bug we're fixing is for trait bounds. That's why the example you tried is correctly throwing you an error already (undefined lifetime). If you try `impl<'a> Copy<'a> for Foo {}` it should throw an error saying that the number of the lifetimes specified is not the one expected.